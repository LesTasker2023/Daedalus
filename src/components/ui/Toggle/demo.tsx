'use client';

import React, { useState } from 'react';
import { Toggle } from '.';

export default function ToggleDemo() {
  const [state, setState] = useState(false);

  return (
    <div className="space-y-6 p-6 max-w-2xl mx-auto">
      <div className="space-y-2">
        <h1 className="text-2xl font-bold">Toggle Component</h1>
        <p className="text-muted-foreground">Toggle button controls</p>
      </div>

      <div className="space-y-4">
        <div className="space-y-2">
          <h3 className="text-lg font-semibold">Basic Usage</h3>
          <div className="border rounded-lg p-4 bg-background">
            <Toggle>Toggle Example</Toggle>
          </div>
        </div>

        <div className="space-y-2">
          <h3 className="text-lg font-semibold">Interactive</h3>
          <div className="border rounded-lg p-4 bg-background space-y-2">
            <Toggle onClick={() => setState(!state)}>
              Click me {state ? '(Active)' : '(Inactive)'}
            </Toggle>
            <p className="text-sm text-muted-foreground">State: {state ? 'Active' : 'Inactive'}</p>
          </div>
        </div>
      </div>
    </div>
  );
}
