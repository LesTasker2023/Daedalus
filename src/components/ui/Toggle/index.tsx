export { default as ToggleDemo } from './demo';
export * from './Toggle';
export type { ToggleProps } from './Toggle';
