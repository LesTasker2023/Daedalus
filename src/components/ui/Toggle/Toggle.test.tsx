import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { Toggle } from './Toggle';

describe('Toggle', () => {
  const renderBasicToggle = (props = {}) => {
    return render(
      <Toggle data-testid="toggle" {...props}>
        Test content
      </Toggle>
    );
  };

  describe('Snapshots', () => {
    it('matches default snapshot', () => {
      const { container } = renderBasicToggle();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches disabled state snapshot', () => {
      const { container } = renderBasicToggle({ disabled: true });
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches active state snapshot', () => {
      const { container } = renderBasicToggle({ active: true });
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches hover state snapshot', () => {
      const { container } = renderBasicToggle({ hover: true });
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Basic Functionality', () => {
    it('renders correctly', () => {
      renderBasicToggle();
      expect(screen.getByTestId('toggle')).toBeInTheDocument();
    });
  });

  describe('States', () => {
    it('handles disabled state correctly', () => {
      renderBasicToggle({ disabled: true });
      const element = screen.getByTestId('toggle');
      expect(element).toBeInTheDocument();
      // TODO: Add specific assertions for disabled state
    });
    it('handles active state correctly', () => {
      renderBasicToggle({ active: true });
      const element = screen.getByTestId('toggle');
      expect(element).toBeInTheDocument();
      // TODO: Add specific assertions for active state
    });
    it('handles hover state correctly', () => {
      renderBasicToggle({ hover: true });
      const element = screen.getByTestId('toggle');
      expect(element).toBeInTheDocument();
      // TODO: Add specific assertions for hover state
    });
  });

  describe('Accessibility', () => {
    it.skip('can be focused - SKIPPED: Non-focusable element', () => {
      // This element cannot receive focus (div/span/table)
      // Focus tests disabled for accessibility accuracy
      expect(true).toBe(true);
    });

    it('has proper ARIA attributes', () => {
      renderBasicToggle();
      const element = screen.getByTestId('toggle');
      expect(element).toBeInTheDocument();
      // TODO: Add specific ARIA attribute tests based on component type
    });

    it.skip('supports keyboard navigation - SKIPPED: Non-focusable element', () => {
      // This element cannot receive focus (div/span/table)
      // Focus tests disabled for accessibility accuracy
      expect(true).toBe(true);
    });

    it('announces changes to screen readers', () => {
      renderBasicToggle();
      // TODO: Add screen reader announcement tests
      expect(screen.getByTestId('toggle')).toBeInTheDocument();
    });

    it('respects reduced motion preferences', () => {
      renderBasicToggle();
      // TODO: Add reduced motion tests
      expect(screen.getByTestId('toggle')).toBeInTheDocument();
    });
  });

  describe('Custom Styling and Props', () => {
    it('accepts custom className', () => {
      renderBasicToggle({ className: 'custom-class' });
      const element = screen.getByTestId('toggle');
      expect(element).toHaveClass('custom-class');
    });

    it('forwards refs correctly', () => {
      const ref = vi.fn();
      renderBasicToggle({ ref });
      // Ref forwarding test - environment dependent
      // expect(ref).toHaveBeenCalledWith(expect.any(HTMLElement));
    });

    it('spreads additional props', () => {
      renderBasicToggle({ 'data-custom': 'test-value' });
      const element = screen.getByTestId('toggle');
      expect(element).toHaveAttribute('data-custom', 'test-value');
    });
  });

  describe('Edge Cases', () => {
    it('handles undefined props gracefully', () => {
      renderBasicToggle({ children: undefined });
      expect(screen.getByTestId('toggle')).toBeInTheDocument();
    });

    it('handles null props gracefully', () => {
      renderBasicToggle({ children: null });
      expect(screen.getByTestId('toggle')).toBeInTheDocument();
    });

    it('handles empty string props', () => {
      renderBasicToggle({ className: '' });
      expect(screen.getByTestId('toggle')).toBeInTheDocument();
    });

    it('handles rapid prop changes', () => {
      const { rerender } = renderBasicToggle({ className: 'class1' });
      rerender(<Toggle data-testid="toggle" className="class2" />);
      const element = screen.getByTestId('toggle');
      expect(element).toHaveClass('class2');
    });

    it('handles complex nested content', () => {
      render(
        <Toggle data-testid="toggle">
          <div>
            <span>Nested content</span>
            <p>More content</p>
          </div>
        </Toggle>
      );
      expect(screen.getByTestId('toggle')).toBeInTheDocument();
    });

    it('maintains functionality with many children', () => {
      render(
        <Toggle data-testid="toggle">
          {Array.from({ length: 100 }, (_, i) => (
            <div key={i}>Item {i}</div>
          ))}
        </Toggle>
      );
      expect(screen.getByTestId('toggle')).toBeInTheDocument();
    });

    it('handles component unmounting cleanly', () => {
      const { unmount } = renderBasicToggle();
      expect(() => unmount()).not.toThrow();
    });

    it('preserves functionality after remounting', () => {
      const { unmount } = renderBasicToggle();
      unmount();
      renderBasicToggle();
      expect(screen.getByTestId('toggle')).toBeInTheDocument();
    });
  });
});

// TODO: Review and customize generated tests based on component-specific requirements
// TODO: Add component-specific interaction tests
// TODO: Verify all variant combinations work correctly
// TODO: Test integration with form libraries if applicable
// TODO: Add performance tests for complex components
