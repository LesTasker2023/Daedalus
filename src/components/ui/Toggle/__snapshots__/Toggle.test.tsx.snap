// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toggle > Snapshots > matches active state snapshot 1`] = `
<button
  aria-pressed="false"
  class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors hover:bg-muted hover:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:[background-color:hsl(var(--surface-interactive))] data-[state=on]:[color:hsl(var(--content-inverse))] bg-transparent h-10 px-3"
  data-state="off"
  data-testid="toggle"
  type="button"
>
  Test content
</button>
`;

exports[`Toggle > Snapshots > matches default snapshot 1`] = `
<button
  aria-pressed="false"
  class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors hover:bg-muted hover:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:[background-color:hsl(var(--surface-interactive))] data-[state=on]:[color:hsl(var(--content-inverse))] bg-transparent h-10 px-3"
  data-state="off"
  data-testid="toggle"
  type="button"
>
  Test content
</button>
`;

exports[`Toggle > Snapshots > matches disabled state snapshot 1`] = `
<button
  aria-pressed="false"
  class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors hover:bg-muted hover:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:[background-color:hsl(var(--surface-interactive))] data-[state=on]:[color:hsl(var(--content-inverse))] bg-transparent h-10 px-3"
  data-disabled=""
  data-state="off"
  data-testid="toggle"
  disabled=""
  type="button"
>
  Test content
</button>
`;

exports[`Toggle > Snapshots > matches hover state snapshot 1`] = `
<button
  aria-pressed="false"
  class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors hover:bg-muted hover:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:[background-color:hsl(var(--surface-interactive))] data-[state=on]:[color:hsl(var(--content-inverse))] bg-transparent h-10 px-3"
  data-state="off"
  data-testid="toggle"
  type="button"
>
  Test content
</button>
`;
